stan/math/rev/core/var.hpp

```diff
@@ -415,6 +416,51 @@ class var_value {
     }
     return os << v.val();
   }
+
+  /**
+   * Assignment of another var value.
+   * @tparam S type of the value in the `var_value` to assing
+   * @param other the value to assign
+   * @return this
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr,
+            require_all_plain_type_t<T, S>* = nullptr>
+  var_value<T> operator=(const var_value<S>& other) {
+    vi_ = other.vi_;
+    return *this;
+  }
+
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr,
+            require_any_not_plain_type_t<T, S>* = nullptr>
+  var_value<T> operator=(const var_value<S>& other) {
+    arena_t<plain_type_t<T>> prev_val = vi_->val_;
+    vi_->val_ = other.val();
+    // no need to change any adjoints - these are just zeros before the reverse
+    // pass
+
+    reverse_pass_callback(
+        [this_vi = this->vi_, other_vi = other.vi_, prev_val]() mutable {
+          this_vi->val_ = prev_val;
+
+          // we have no way of detecting aliasing between this->vi_->adj_ and
+          // other.vi_->adj_, so we must copy adjoint before reseting to zero
+
+          // we can reuse prev_val instead of allocating a new matrix
+          prev_val = this_vi->adj_;
+          this_vi->adj_.setZero();
+          other_vi->adj_ += prev_val;
```

Comment:
Should we reset `this_vi->adj_` to `prev_val` here?

Reply:
No. First at this point in code `prev_val` contains the value of `this_vi->adj_`. From this reverse pass callback on `this_vi->adj_` should contain adjoint associated with the values in this block before they were overwritten. Since they were overwritten, they were not used in any other operations and their adjoint is 0, which is set on line 450.